action/protocol/staking/bucket_pool.go

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2020 IoTeX Foundation
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package staking
+
+import (
+	"context"
+	"github.com/pkg/errors"
+	"math/big"
+
+	"github.com/iotexproject/go-pkgs/hash"
+	"github.com/iotexproject/iotex-address/address"
+
+	"github.com/iotexproject/iotex-core/action/protocol"
+	accountutil "github.com/iotexproject/iotex-core/action/protocol/account/util"
+	"github.com/iotexproject/iotex-core/config"
+	"github.com/iotexproject/iotex-core/state"
+)
+
+// when a bucket is created, the amount of staked IOTX token is deducted from user, but does not transfer to any address
+// in the same way when a bucket are withdrawn, bucket amount is added back to user, but does not come out of any address
+//
+// for better accounting/auditing, we take protocol's address as the 'bucket pool' address
+// 1. at Greenland height we sum up all existing bucket's amount and set the total amount to bucket pool address
+// 2. for future bucket creation/deposit/registration, the amount of staked IOTX token will be added to bucket pool (so
+//    the pool is 'receiving' token)
+// 3. for future bucket withdrawal, the bucket amount will be deducted from bucket pool (so the pool is 'releasing' token)
+
+type (
+	bucketPool struct {
+		addr  address.Address
+		pool  *state.Account
+		hu    config.HeightUpgrade
+		exist bool
+		total *big.Int
+	}
+)
+
+func NewBucketPool(addr address.Address) *bucketPool {
+	s := state.EmptyAccount()
+	return &bucketPool{
+		addr:  addr,
+		pool:  &s,
+		total: big.NewInt(0),
```

Comment:
need implement readstate method: https://github.com/iotexproject/iotex-proto/blob/master/proto/api/read_state.proto#L66
 to read out total amount with corresponding height

Reply:
yes, this is a TODO

Reply:
done in latest commit